### ISSUE TITLE
ClientWebSocket.ConnectAsync() does not work on Mac 

### ISSUE THREAD

**Issue by linkaiyu** [dotnet/runtime#90258]
### Description

in a simple MAUI test app, call to `ClientWebSocket.ConnectAsync()` on Mac does not return. 
some experimental work suggests that MAUI is defaulting to `HttpNoProxy`:
when we don’t set proxy to `ClientWebSocket` Option the MAUI `ClientWebSocket` is defaulting to `NoProxy` and the connect fails. I checked the values of `HttpClient.DefaultProxy`:
In MAUI app:
```c#
  HttpClient.DefaultProxy    {System.Net.Http.HttpNoProxy}
```

comparing that to .Net7 App, it's set to `MacProxy`:
```c#
  HttpClient.DefaultProxy    {System.Net.Http.MacProxy}  
```
for reference:
https://github.com/dotnet/dotnet/blob/main/src/runtime/src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L35

in customer situation, WPAD (Web Proxy Auto-Discovery) is used to configure internal laptops. 
while testing on MS VPN, the code works. It's worth considering whether the MS VPN proxy setup differs from WPAD, 

we need to get Proxy Url/object into MAUI app for all different VPN configuration and create `WebProxy` object and set it to `ClientWebSocket`.

### Steps to Reproduce

download the repro to Mac connected via VPN with WPAD configured.
in the repro app, click on button to start test. the call to `ClientWebSocket.ConnectAsync()` will be stuck 

### Link to public reproduction project repository

https://github.com/linkaiyu/ClientWebSocket

### Version with bug

7.0.49

### Last version that worked well

Unknown/Other

### Affected platforms

macOS

### Did you find any workaround?

short term workaround is to create a webproxy for the connection, but for long term when the app is going to be released to the public, the short term solution won't meet the needs

**Comment by CarnaViire** [comment]
From the issue:

> I checked the values of HttpClient.DefaultProxy:
> In MAUI app:
>HttpClient.DefaultProxy {System.Net.Http.HttpNoProxy}
>
> comparing that to .Net7 App, it's set to MacProxy:
>HttpClient.DefaultProxy {System.Net.Http.MacProxy}

This doesn't look like a runtime issue, since it seems like either MAUI or something else is overriding the defaults somehow. @drasticactions @samhouts @linkaiyu

From what I see, on macOS, iOS and tvOS we use `MacProxy` by default:

https://github.com/dotnet/runtime/blob/943af3a6fe4300d1722d9ad6c9b371c568bc9fe4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SystemProxyInfo.OSX.cs#L11

<details>

https://github.com/dotnet/runtime/blob/943af3a6fe4300d1722d9ad6c9b371c568bc9fe4/src/libraries/System.Net.Http/src/System.Net.Http.csproj#L323-L325

</details>

while on other Unixes we use `HttpNoProxy`:

https://github.com/dotnet/runtime/blob/943af3a6fe4300d1722d9ad6c9b371c568bc9fe4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SystemProxyInfo.Unix.cs#L13

<details>

https://github.com/dotnet/runtime/blob/943af3a6fe4300d1722d9ad6c9b371c568bc9fe4/src/libraries/System.Net.Http/src/System.Net.Http.csproj#L319-L321

</details>

It also shows that it can be overridden by an environmental value, so it is another potential culprit.

**Comment by CarnaViire** [comment]
cc @wfurt @MihaZupan

**Comment by wfurt** [comment]
It would be good to get packet captures @linkaiyu and verify where this code is trying to connect to. 
AFAIK the discovery does not work on macOS e.g. you will either need to set the environment or use the system proxy config. It may be best to try very simple project with HttpCLient to peel off all the extra layers.

**Comment by linkaiyu** [comment]
customer experimented with pre .net7 app and it works.
the problem happens with MAUI app.
will reach out to customer for network trace, not sure if they can provide useful info.

@drasticactions @CarnaViire @wfurt you are all added access to the repro app. very simple code, just a button click to invoke the API.

**Comment by linkaiyu** [comment]
@wfurt , customer captured pcap file
https://github.com/linkaiyu/ClientWebSocket/issues

it contains two use cases. 
the first one is the MAUI app makes ClientWebSocket call that results in call never return
the second one is the .NET app makes ClientWebSocket call that works fine.

We compared the customer's Mac machine with Jon Goldber's Mac machine. The different is that customer is on VPN with WPAD. Jon's machine is on MSFT VPN without WPAD. 

So the MAUI app running on VPN with WPAD seems to be the issue.

**Comment by linkaiyu** [comment]
@karelz , the console app works. the customer repro include a console app that uses the same code as in MAUI app. MAUI app fails.
@mikeparker104

**Comment by wfurt** [comment]
I don't have WPAD setup at the moment but could you flip `System.Net.Http.UseNativeHttpHandler` to `false`  in `ClientWebSocket.Maui.runtimeconfig.json` (or what ever runtime config) and report results @linkaiyu ? 

it seems like when it is build for MAUI it targets `maccatalyst` and that pulls in native handler as far as I can tell. 
@steveisok would probably have more insight. 

There maybe another workaround with something like this in startup:
```c#
var handler = new SocketsHttpHandler();
HttpClient.DefaultProxy = handler.Proxy;
```

**Comment by karelz** [comment]
@linkaiyu thanks for trying it out!
If the console app works, then it is most likely difference in the handlers in MAUI vs. on Mac (i.e. the workaround @wfurt suggested is likely going to help).
Either way, @steveisok's team has expertise in native handlers and he will take further look - I transferred the investigation to him.
Thanks!
cc @marek-safar

**Comment by steveisok** [comment]
@linkaiyu can I get access to the repo you linked in the issue?

**Comment by steveisok** [comment]
The reason why it's defaulting to `System.Net.Http.HttpNoProxy` is due to a missing condition for maccatlyst when including the source file.
 https://github.com/dotnet/runtime/blob/943af3a6fe4300d1722d9ad6c9b371c568bc9fe4/src/libraries/System.Net.Http/src/System.Net.Http.csproj#L319-L325

That will also prevent SocketsHttpHandler from working properly if you were to set `<UseNativeHttpHandler>` to false.

@mandel-macaque has some knowledge of how the native handler (NSUrlSessionHandler) can be configured to pick up the auto config of the system.

**Comment by linkaiyu** [comment]
Thanks Steve.

done.

Regards,
Linkai
________________________________

**Comment by linkaiyu** [comment]
@mandel-macaque you are already added to the access list, please accept the invite
Thanks
Linkai

**Comment by linkaiyu** [comment]
@mandel-macaque 
Hi Mandel, can you confirm if you have added the workaround code to the customer repro please?  I will have customer test it once it's updated.
Thanks
Linkai
@mikeparker104

**Comment by linkaiyu** [comment]
> The simplest way to workaround this for the user is to create a NSUrlSessionHandler that takes a configuration that allows the native handler use the automatic handler configuration. The way to do so it is as follows:
> 
> ## Create a default configuration and update its settings
> [NSUrlSessionConfiguration](https://learn.microsoft.com/en-us/dotnet/api/foundation.nsurlsessionconfiguration?view=xamarin-ios-sdk-12) does not have public constructors, the appropriate way to create a configuration is to request the default configuration and update its settings. You can do so by using the property [DefaultSessionConfiguration](https://learn.microsoft.com/en-us/dotnet/api/foundation.nsurlsessionconfiguration.defaultsessionconfiguration?view=xamarin-ios-sdk-12#foundation-nsurlsessionconfiguration-defaultsessionconfiguration), once you have the default session configuration you have to update the proxy settings via the [ConnectionProxyDictionary](https://learn.microsoft.com/en-us/dotnet/api/foundation.nsurlsessionconfiguration.connectionproxydictionary?view=xamarin-ios-sdk-12#foundation-nsurlsessionconfiguration-connectionproxydictionary). In that dictionary you have to use the key 'kCFNetworkProxiesProxyAutoConfigEnable', unfortunately they key is not public but in my sample code I show a way to retrieve it.
> 
> ### Retrieve key value
> ```cs
> using ObjCRuntime;
> 
> static NSString? kCFProxyAutoConfigurationHTTPResponseKey;
> static NSString? AutoConfigurationHTTPResponseKey {
>     get {
>         if (kCFProxyAutoConfigurationHTTPResponseKey is null)
>             kCFProxyAutoConfigurationHTTPResponseKey = Dlfcn.GetStringConstant (Libraries.CFNetwork.Handle, "kCFProxyAutoConfigurationHTTPResponseKey");
> 
>             return kCFProxyAutoConfigurationHTTPResponseKey;
>      }
> }
> ```
> 
> ### Create config
> ```cs
> using var configuration = NSUrlSessionConfiguration.DefaultSessionConfiguration;
> 
> // IMPORTANT: set the timeouts, this is done by Microsoft.iOS for you, but has to be maintained once you get the default one
> // because we want the timeout from HttpClient to have precedence over the one from NSUrlSession
> // Double.MaxValue does not work, so default to 24 hours
> 
> configuration.TimeoutIntervalForRequest = 24 * 60 * 60;
> configuration.TimeoutIntervalForResource = 24 * 60 * 60;
> 
> 
> // set the config to use the default auto config proxies:
> NSObject[] values = new NSObject[]
> {
>     NSNumber.FromInt32 (1)                   //Enable auto proxy
> 
> };
> 
> NSObject[] keys = new NSObject[]
> {
>     AutoConfigurationHTTPResponseKey
> };
> 
> NSDictionary proxyDict = NSDictionary.FromObjectsAndKeys (values, keys);
> configuration.ConnectionProxyDictionary = proxyDict;
> ```
> 
> ## Create the NSUrlSessionHandler
> Once you have the configuration properly set, you can create a NSUrlSessionHandler with it and pass it to the http client:
> 
> ```cs
> var handler =  NSUrlSessionHandler (configuration);
> var httpClient = new HttpClient (handler);
> ```
> 
> ## Notes
> * All of the native objects are IDisposable, please make sure that the memory management in the application does consider this.
> * I have created an issue in the iOS repo to expose those keys: https://github.com/xamarin/xamarin-macios/issues/18911
> * We might want to update the NSUrlSessionHandler API to consider this uses cases.
> 
> If you give me access to the repo I will add the appropriate code there so that we can test their settings and that way they have a working example.

@mandel-macaque , customer tried to compile but got these errors:

Error CS0122: 'Libraries' is inaccessible due to its protection level (CS0122)
Error CS1955: Non-invocable member 'NSUrlSessionHandler' cannot be used like a method. (CS1955) (ClientWebSocket.Maui)

**Comment by linkaiyu** [comment]
Hi @steveisok , @mandel-macaque tried to compile with his workaround code but failed, customer also had the compilation error. As @mandel-macaque suggested, the workload with your updated network proj needs to be updated. We can advise customer to wait till the next SR. To avoid any unpredicted "gotcha", can you please pull down @mandel-macaque 's workaround code and test compiling on your machine (I assume you have the latest workload build for Mac)?

@mandel-macaque , can you please merge your workaround code into the repro please so that Steve can test it?
Thanks

**Comment by linkaiyu** [comment]
@mandel-macaque , customer tested the repro app with your code but was still seeing the hang issue. here is what he told me:

"I am still getting hanging issue while connecting to streaming server testing with fidelity internal laptop.
Code initially was setting null to web proxy and I tried passing the MacProxy it did not help.

IWebProxy webProxy = new MacProxy();

_clientWebSocketService.ConnectAsync(webProxy).Wait();
"

**Comment by weedcry** [comment]
@mandel-macaque 
I am using a proxy for NSUrlSessionHandler in the following way:
Refer to this discussion [86364](https://developer.apple.com/forums/thread/86364)

```
var configuration = NSUrlSessionConfiguration.DefaultSessionConfiguration;

var dictionary = new NSDictionary(
    new NSString("HTTPEnable"), 1,
    new NSString("HTTPProxy"), ProxyServer,
    new NSString("HTTPPort"), ProxyPort,
    new NSString("HTTPSEnable"), 1,
    new NSString("HTTPSProxy"), ProxyServer,
    new NSString("HTTPSPort"), ProxyPort);

configuration.ConnectionProxyDictionary = dictionary;
var handler =  NSUrlSessionHandler(configuration);
var httpClient = new HttpClient (handler);
```
Using this method, I have been able to send requests through the proxy server.
Do you think this method is okay? Will there be any impact when the proxy setting of NSUrlSessionHandler is not supported?

**Comment by mandel-macaque** [comment]
@weedcry  sorry for the late reply, I was on PTO. That is a way to set the proxy to be used just remember that it is ignoring the default configurations. But it is a perfectly valid way to do this.

**Comment by linkaiyu** [comment]
Yes

Thanks
Linkai


________________________________

**Comment by karelz** [comment]
Fixed in main (9.0) in PR #91473 and in 8.0.0 (RC2) in PR #91599 and in 7.0.13 in PR #91817.
